runtime/libc/wasi_sdk_backing.c

```diff
@@ -0,0 +1,339 @@
+        flags |= O_RDONLY;
+    }
+
+    int mode = 0644;
+    int fd = openat(dirfd, path, flags, mode);
+    if (fd < 0) {
+        return -errno;
+    }
+
+    set_i32(fd_off, fd);
+    return 0;
+}
+
+i32 wasi_unstable_fd_close(i32 fd) {
+    i32 res = (i32) close(fd);
+
+    if (res == -1) {
+        return -errno;
+    }
+    return res;
+}
+
+i32 wasi_unstable_fd_fdstat_get(i32 fd, u32 buf_offset) {
+    struct wasi_fdstat* fdstat = get_memory_ptr_void(buf_offset, sizeof(struct wasi_fdstat));
+
+    struct stat stat;
+    i32 res = fstat(fd, &stat);
+    if (res == -1) {
+        return -errno;
+    }
+    int mode = stat.st_mode;
+
+    i32 fl = fcntl(fd, F_GETFL);
+    if (fl < 0) {
+        return -errno;
+    }
+
+    fdstat->fs_filetype = (
+            (S_ISBLK(mode)  ? WASI_FILETYPE_BLOCK_DEVICE     : 0) |
+            (S_ISCHR(mode)  ? WASI_FILETYPE_CHARACTER_DEVICE : 0) |
+            (S_ISDIR(mode)  ? WASI_FILETYPE_DIRECTORY        : 0) |
+            (S_ISREG(mode)  ? WASI_FILETYPE_REGULAR_FILE     : 0) |
+            (S_ISSOCK(mode) ? WASI_FILETYPE_SOCKET_STREAM    : 0) |
+            (S_ISLNK(mode)  ? WASI_FILETYPE_SYMBOLIC_LINK    : 0));
+    fdstat->fs_flags = (
+            ((fl & O_APPEND)   ? WASI_FDFLAG_APPEND     : 0) |
+            ((fl & O_DSYNC)    ? WASI_FDFLAG_DSYNC      : 0) |
+            ((fl & O_NONBLOCK) ? WASI_FDFLAG_NONBLOCK   : 0) |
+            ((fl & O_RSYNC)    ? WASI_FDFLAG_RSYNC      : 0) |
+            ((fl & O_SYNC)     ? WASI_FDFLAG_SYNC       : 0));
+    fdstat->fs_rights_base = 0; // all rights
+    fdstat->fs_rights_inheriting = 0; // all rights
+
+    return 0;
+}
+
+i32 wasi_unstable_fd_fdstat_set_flags(i32 fd, u32 fdflags) {
+    int flags = (
+        ((flags & WASI_FDFLAG_APPEND  ) ? O_APPEND   : 0) |
+        ((flags & WASI_FDFLAG_DSYNC   ) ? O_DSYNC    : 0) |
+        ((flags & WASI_FDFLAG_NONBLOCK) ? O_NONBLOCK : 0) |
+        ((flags & WASI_FDFLAG_RSYNC   ) ? O_RSYNC    : 0) |
+        ((flags & WASI_FDFLAG_SYNC    ) ? O_SYNC     : 0));
+    int err = fcntl(fd, F_SETFL, fdflags);
+    if (err < 0) {
+        return -errno;
+    }
+    return 0;
+}
+
+i32 wasi_unstable_fd_seek(i32 fd, i64 file_offset, i32 whence, u32 newoffset_off) {
+    off_t res = lseek(fd, (off_t)file_offset, whence);
+
+    if (res == -1) {
+        return -errno;
+    }
+
+    set_i64(newoffset_off, res);
+    return 0;
+}
+
+i32 wasi_unstable_fd_datasync(i32 fd) {
+    int res = fdatasync(fd);
+    if (res == -1) {
+        return -errno;
+    }
+
+    return 0;
+}
+
+i32 wasi_unstable_fd_read(i32 fd, i32 iov_offset, i32 iovcnt, i32 nread_off) {
+    i32 sum = 0;
+    struct wasi_iovec *iov = get_memory_ptr_void(iov_offset, iovcnt * sizeof(struct wasi_iovec));
+
+    for (int i = 0; i < iovcnt; i++) {
+        void* ptr = get_memory_ptr_void(iov[i].base_offset, iov[i].len);
+        ssize_t res = read(fd, ptr, iov[i].len);
+        if (res == -1) {
+            return -errno;
```

Comment:
I don't know if WASI errnos map 1:1 to POSIX. Hope so!

Reply:
Hmm, initially I just carried over the errno handling from the wasmception_backing.c. But the implying the WASI API may mean these do need use a stricter conversion. I will update. It would get optimized out if they are identical anyways.